### DISCUSSION TITLE
Unable to return results for both Sharepoint and Azure

### DISCUSSION THREAD

**Discussion by garretlayel** [silverhack/monkey365#14] [Q&A]
When running the example scripts, multiple errors and warning messages occur. They explain that query results did not return any data inside of the tenant. This occurs for dozens of lines in the query results. In addition, for SharePoint, the error messages explain that the script is unable to get site properties for the SharePoint account. I have correct permissions in both Azure and SharePoint. Am I running the script wrong? I previously unblocked and imported the monkey365 module; however, no results populate except for warning/error messages. Any help is appreciated. Thanks!

**Comment by silverhack** [comment]
Hi, it seems you’re executing Monkey365 with incorrect permissions. Which permission are you using? Please note that for SharePoint Online, you will need grant the user the role of SharePoint Administrator.

https://silverhack.github.io/monkey365/permissions/

**Reply by garretlayel** [reply]
As of now and when I tried the script, I have global reader, security reader, and SharePoint admin. I am also running PowerShell as an admin, so the execution is authorized

**Comment by silverhack** [comment]
Hi! I can’t reproduce the issue in both Azure and SharePoint. Could you please check if that user has access to the SharePoint site?

**Reply by garretlayel** [reply]
Yes, I will reach out regarding access. Thank you, I will remain in contact when I find out!

**Reply by silverhack** [reply]
Cool thanks!

**Reply by garretlayel** [reply]
I got the script to work correctly. But is there a way to share the HTML file with coworkers? Right now, I can send the file but it does not open correctly from their point of view.

**Comment by davevh007** [comment]
Hello,

First off, thank you for creating and maintaining this tool!

I last ran the report around mid September with the Monkey365 version at the time and the O365 data was retrieved as expected (using Powershell 5).

However, I tried running the report today with the latest version from your repo (updated 2 days ago for 2 file changes) and all data is retrieved except Onedrive/Sharepoint. I am using the same script syntax as before except for the new instance name change as below:
<
Invoke-Monkey365  -PromptBehavior SelectAccount -IncludeAzureActiveDirectory -Instance Microsoft365 -Analysis MicrosoftTeams,SharePointOnline,ExchangeOnline,PurView,IRM,MicrosoftForms -ExportTo html
>

There are no errors, just no data being returned for SPO/Onedrive. All permissions are correct, I have tried Global Admin/Sharepoint Admin to no avail. Tried through both PS V5 and V7. I also tried the Monkey365 version I used in September using the same syntax as above and this time it also failed to retrive the SPO data.

Any advice?

Thanks!

**Reply by silverhack** [reply]
Hello @davevh007 hope you're doing well and thanks for reporting this. 

I can reproduce your issue when copying and pasting example commands from docs. The internal function that manages plugins is case sensitive, so is returning an empty array if the string does not match. As a workaround, you could run the following command (Please, note the lack of capital letter in SharePointOnline):

```
Invoke-Monkey365 -PromptBehavior SelectAccount -IncludeAzureActiveDirectory -Instance Microsoft365 -Analysis MicrosoftTeams,SharepointOnline,ExchangeOnline,PurView,IRM,MicrosoftForms -ExportTo html
```
As a recommendation, and in order to avoid misspelling commands, I'd recommend to use of [TAB-completion](https://learn.microsoft.com/en-us/powershell/scripting/learn/shell/tab-completion?view=powershell-7.2) in PowerShell.

On the other hand, I'll open a new issue to track it. 

Thanks!

**Reply by silverhack** [reply]
Hello @davevh007, this is now fixed in both main and develop branches. 

Cheers,

**Reply by davevh007** [reply]
@silverhack  Appreciate the quick response and code push.

However, I ran with the case recommended (and with the latest branch update) and same issue (I ran from Visual Studio code under PS7, directly from PS7 to rule out any VS Code issues, and from PS5 in the IDE all elevated to Admin) occured, no SPO data (FYI VS Code does suggest case change for SPO to "SharepointOnline" so appreciate the tip). Here is the initial output from PS5:

"
The property 'Message' cannot be found on this object. Verify that the property exists.
At C:\Program Files\WindowsPowerShell\Modules\monkey365\core\modules\monkeycloudutils\public\Get-TenantInfo.ps1:61 char:12
+         if($_.ErrorDetails.Message){
+            ~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], PropertyNotFoundException
    + FullyQualifiedErrorId : PropertyNotFoundStrict
 
WARNING: Unable to get token for SharePoint Online
The property 'Message' cannot be found on this object. Verify that the property exists.
At C:\Program Files\WindowsPowerShell\Modules\monkey365\core\modules\monkeycloudutils\public\Get-TenantInfo.ps1:61 char:12
+         if($_.ErrorDetails.Message){
+            ~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], PropertyNotFoundException
    + FullyQualifiedErrorId : PropertyNotFoundStrict
 
WARNING: Unable to get token for SharePoint Online
The property 'Message' cannot be found on this object. Verify that the property exists.
At C:\Program Files\WindowsPowerShell\Modules\monkey365\core\modules\monkeycloudutils\public\Get-TenantInfo.ps1:61 char:12
+         if($_.ErrorDetails.Message){
+            ~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], PropertyNotFoundException
    + FullyQualifiedErrorId : PropertyNotFoundStrict
"

I believe I received that same output when I ran the script before and it pulled the SPO data. Below is the full output running under PS7, identifying data replaced with XXX (Azure permissions elevated to Global Admin/SPO Admin):

"
WARNING: No account was found on public application
WARNING: Your connection has been redirected to the following URI: "https://nam04b.ps.compliance.protection.outlook.com/Powershell-LiveId?BasicAuthToOAuthConversion=true;PSVersion=7.2.7"
WARNING: [13:22:16:984] - [Get-MonkeyEXOPhishFilterPolicy] - The Exchange Phish filter policy query did not return any data in XXX tenant - warning - XXX.XXX.XXX - ExoPhishFilterResponse
WARNING: [13:22:20:813] - [Get-MonkeyEXOAuthenticationPolicy] - The Exchange Online authentication policy query did not return any data in XXX tenant - warning - XXX.XXX.XXX - ExoAuthPolicyEmptyResponse
WARNING: [13:22:24:151] - [Get-MonkeyEXOHostedContentFilterRule] - The Exchange Online Hosted content filter rule query did not return any data in XXX tenant - warning - XXX.XXX.XXX - ExoHostedContentEmptyResponse
WARNING: [13:22:27:472] - [Get-MonkeyEXODevicePolicy] - The Exchange Online device policy query did not return any data in XXX tenant - warning - XXX.XXX.XXX - ExoDevicePolicyEmptyResponse
WARNING: [13:22:37:311] - [Get-MonkeyEXOMalwareFilterRule] - The Exchange Online malware filter rule query did not return any data in XXX tenant - warning - XXX.XXX.XXX - ExoMalwareFilterRuleEmptyResponse
WARNING: [13:25:55:727] - [Get-MonkeyEXOActivityAlert] - The Security and Compliance activity alerts query did not return any data in XXX tenant - warning - XXX.XXX.XXX - SecCompActivityAlertsEmptyResponse
WARNING: [13:26:11:553] - [Get-MonkeyEXOLabelPolicy] - The Security and Compliance label policy query did not return any data in XXX tenant - warning - XXX.XXX.XXX - SecCompLabelPolicyEmptyResponse
WARNING: [13:28:47:919] - [Get-MonkeyADPasswordPolicy] - The Azure AD password template policy query did not return any data in XXX tenant - warning - XXX.XXX.XXX - AzurePortalTemplatePolicyEmptyResponse
WARNING: [13:28:54:546] - [Get-MonkeyADPortalDeviceSetting] - The Azure AD device settings query did not return any data in XXX tenant - warning - XXX.XXX.XXX - AzurePortalDevicesEmptyResponse
WARNING: [13:28:57:891] - [Get-MonkeyADRoamingInfo] - The Azure AD Roaming properties query did not return any data in XXX tenant - warning - XXX.XXX.XXX - AzurePortalRoamingEmptyResponse
WARNING: [13:29:09:740] - [Get-MonkeyTeamsNotificationsFeedsPolicy] - The Microsoft 365 Teams: Notifications and feeds policy query did not return any data in XXX tenant - warning - XXX.XXX.XX - TeamsOrgSettingsEmptyResponse
WARNING: [13:29:09:815] - [Get-MonkeyTeamsOrgConfiguration] - The Microsoft 365 Teams: Organisation settings query did not return any data in XXX tenant - warning - XXX.XXX.XXX - TeamsOrgSettingsEmptyResponse
WARNING: [13:29:10:325] - [Get-MonkeyTeamsTargetingPolicy] - The Microsoft 365 Teams: Targeting policy query did not return any data in XXX tenant - warning - XXX.XXX.XXX - TeamsTargetingPolicyEmptyResponse

Environment                   : {AzurePortal, Graphv2, ExchangeOnline, WebAppServicePortal…}
internal_modules              : {core/modules/monkeylogger, core/modules/monkeywebrequest, core/modules/monkeyast, core/modules/monkeyjob…}
msal_modules                  : {core/modules/monkeycloudutils, core/modules/monkeymsal, core/modules/monkeymsalauthassistant}
adal_modules                  : {core/modules/monkeycloudutils, core/modules/monkeyadal, core/modules/monkeyadalauthassistant}
watcher                       : {core/watcher}
runspaces_modules             : {C:\Program Files\WindowsPowerShell\Modules\monkey365/core/modules/monkeywebrequest, C:\Program Files\WindowsPowerShell\Modules\monkey365/core/modules/monkeylogger, C:\Program Files\WindowsPowerShell\Modules\monkey365/core/modules/monkeyast, C:\Program 
                                Files\WindowsPowerShell\Modules\monkey365/core/modules/monkeyjob…}
runspace_init                 : {C:\Program Files\WindowsPowerShell\Modules\monkey365/core/runspace_init/Initialize-MonkeyRunspace.ps1}
exo_runspace_init             : {C:\Program Files\WindowsPowerShell\Modules\monkey365/core/runspace_init/Initialize-MonkeyExoRunspace.ps1}
libutils                      : {C:\Program Files\WindowsPowerShell\Modules\monkey365\core\api\azure\graph\helpers\general\Get-MonkeyADObjectByObjectId.ps1, C:\Program Files\WindowsPowerShell\Modules\monkey365\core\api\azure\graph\helpers\group\Get-MonkeyAZGroupMember.ps1, C:\Program 
                                Files\WindowsPowerShell\Modules\monkey365\core\api\azure\graph\helpers\rbac\Group-MonkeyAZRBACMember.ps1, C:\Program Files\WindowsPowerShell\Modules\monkey365\core\api\azure\graph\helpers\user\Get-AADDetailedUser.ps1…}
OnlineServices                : {O365, Teams, SPS, Forms…}
Localpath                     : C:\Program Files\WindowsPowerShell\Modules\monkey365
auth_tokens                   : {AzurePortal, AzureStorage, Intune, ExchangeOnline…}
userPrincipalName             : XXX
userId                        : XXX
orgRegions                    : XXX
Tenant                        : @{SKU=System.Object[]; CompanyInfo=; TenantId=XXX; Domains=System.Object[]; MyDomain=; TenantName=XXX}
Plugins                       : {@{Id=aad0001; Provider=AzureAD; Title=Plugin to get azure apps from Azure AD; Group=System.String[]; ServiceName=Azure AD Applications; PluginName=Get-MonkeyADApplication; Docs=https://silverhack.github.io/monkey365/; File=C:\Program
                                Files\WindowsPowerShell\Modules\monkey365\plugins\aad\portal\applications\Get-MonkeyADApplication.ps1}, @{Id=aad0002; Provider=AzureAD; Title=Plugin to get audit logs from Azure AD; Group=System.String[]; ServiceName=Azure AD Audit; PluginName=Get-MonkeyADAudit;     
                                Docs=https://silverhack.github.io/monkey365/; File=C:\Program Files\WindowsPowerShell\Modules\monkey365\plugins\aad\portal\audit\Get-MonkeyADAudit.ps1}, @{Id=aad0003; Provider=AzureAD; Title=Plugin to get password reset policy from Azure AD; Group=System.String[];   
                                ServiceName=Azure AD SSPR; PluginName=Get-MonkeyADAuthorisationPolicy; Docs=https://silverhack.github.io/monkey365/; File=C:\Program Files\WindowsPowerShell\Modules\monkey365\plugins\aad\portal\authorisation\Get-MonkeyADAuthorisationPolicy.ps1}, @{Id=aad0005;        
                                Provider=AzureAD; Title=Plugin to get Directoryroles from Azure AD; Group=System.String[]; ServiceName=Azure AD Directory Role; PluginName=Get-MonkeyADPortalDirectoryRoleInfo; Docs=https://silverhack.github.io/monkey365/; File=C:\Program
                                Files\WindowsPowerShell\Modules\monkey365\plugins\aad\portal\directoryroles\Get-MonkeyADPortalDirectoryRoleInfo.ps1}…}
ATPEnabled                    : True
Licensing                     : {@{capabilityStatus=Enabled; consumedUnits=21; objectId=XXX; prepaidUnits=; servicePlans=System.Object[]; skuId=XXX; skuPartNumber=XXX; appliesTo=User},    
                                @{capabilityStatus=Enabled; consumedUnits=139; objectId=XXX; prepaidUnits=; servicePlans=System.Object[]; skuId=XXX; skuPartNumber=XXX; appliesTo=User},   
                                @{capabilityStatus=Enabled; consumedUnits=141; objectId=XXX; prepaidUnits=; servicePlans=System.Object[]; skuId=XXX; skuPartNumber=XXX; appliesTo=User},     
                                @{capabilityStatus=Enabled; consumedUnits=2; objectId=XXX; prepaidUnits=; servicePlans=System.Object[]; skuId=XXX; skuPartNumber=XXX;
                                appliesTo=User}…}
LogPath                       : 
exo_session_start_time        : 
compliance_session_start_time : 
exo_msal_application          : Microsoft.Identity.Client.PublicClientApplication
sps_msal_application          : 
authContext                   : 
msalapplication               : Microsoft.Identity.Client.PublicClientApplication
adal_credentials              : 
adal_application              : 
application_args              : {Environment, InformationAction, Debug, Verbose}
msal_application_args         : {publicApp, InformationAction, Silent, PromptBehavior…}
msal_client_app_args          : 
o365_sessions                 : {}
isConfidentialApp             : False
isUsingAdalLib                : 
InformationAction             : SilentlyContinue
VerboseOptions                : {Verbose, Debug}
AuthType                      : Interactive
WriteLog                      : 
userAgent                     : 
subscriptions                 : 
current_subscription          : {Tenant, DisplayName, subscriptionId}
userPermissions               : @{user=; userPrincipalName=XXX; displayName=XXX; permissions=System.Object[]}
all_resources                 : 
ResourceGroups                :
internal_config               : @{title=Monkey 365 Config File; requirements=; version=; debug=; httpSettings=; jsonOptions=; performance=; profileSettings=; htmlSettings=; ruleSettings=; logging=; excelSettings=; o365=; azuread=; resourceManager=System.Object[]}
dlp                           : {@{Name=DLP Policies for Government Data; Tag=Government_Data; Sits=System.Object[]}, @{Name=DLP Policies for Company Sensitive Information; Tag=Company_Sensitive_Information; Sits=System.Object[]}, @{Name=DLP Policies for ePHI; Tag=Health_Information;
                                Sits=System.Object[]}, @{Name=DLP Policies for Personally Identifiable Information; Tag=PII; Sits=System.Object[]}…}
AuthResponses                 :
executionInfo                 : @{Domain=XXX; userPrincipalName=XXX; displayName=XXX; permissions=; roles=; ScanDate=10/28/2022 8:20:43 AM; subscription=System.Collections.Specialized.OrderedDictionary; tenant=; userpic=XXX 
                                XXX 
                                XXX 
                                XXX 
                                XXX 
                                XXX 
                                XXX==; profile=System.Collections.Specialized.OrderedDictionary}        
startDate                     : 10/28/2022 8:20:43 AM
exo_runspacePool              :
monkey_runspacePool           :
threads                       : 2
SaveProject                   :
Instance                      : Microsoft365
IncludeAAD                    : True
verbose                       : False
debug                         : False
initParams                    : {[PromptBehavior, SelectAccount], [IncludeAzureActiveDirectory, True], [Instance, Microsoft365], [ExportTo, System.Object[]]…}
clientApplicationId           :
TenantId                      : XXX
exportTo                      : {html}
LocalizedDataParams           : {FileName, BaseDirectory, BindingVariable}
BatchSleep                    : 4000
BatchSize                     : 200
MaxQueue                      : 2
nestedRunspaceMaxThreads      : 1
outDir                        : C:\Program Files\WindowsPowerShell\Modules\monkey365/monkey-reports
excludePlugins                :
excludedResources             :
"

It pulls all the other data correctly:
![image](https://user-images.githubusercontent.com/87340378/198668893-9bdb7b70-3c1a-46c8-83dc-ff13a98f44c9.png)

**Reply by silverhack** [reply]
Hey @davevh007, 

It seems that Monkey365 was unable to get the token for SharePointOnline, probably due to a timeout. And for that reason you'll get a report without SharePoint & OneDrive data.

Could you please test again? I've tested the latest release with multiple tenants and it works as expected. 

On the other hand, Monkey365 should not load plugins if connection is not returned, so I'll add a new issue to track this enhancement. 

Thanks!

**Reply by davevh007** [reply]
@silverhack I will give it another pass next week, enjoy your weekend!

**Reply by davevh007** [reply]
@silverhack I gave it another pass today on a completely fresh installation on another machine and still unable to pull any SPO data. I think this may have something to do with Microsoft doing migrations from the lagacy portal or something specific to our tenant. No matter, from what I can tell the report from a month ago when it succeeded in pulling SPO data is still valid.

Thanks!
